tools/project_manager/page_compare.php

```diff
@@ -16,18 +16,25 @@ class Comparator
 {
     public function __construct()
     {
-        global $PROJECT_STATES_IN_ORDER;
-
-        $this->L_round_options = ['P1', 'P2', 'P3', 'F1'];
-        $this->R_round_options = ['F1', 'F2'];
-        $this->state_index = array_flip($PROJECT_STATES_IN_ORDER);
+        $all_rounds = Rounds::get_all();
+        $all_ids = Rounds::get_ids();
+
+        // L_round_options are all rounds except the last
+        $this->L_round_options = array_slice($all_ids, 0, count($all_ids) - 1);
+        // R_round_options are all formatting rounds
+        $this->R_round_options = array_keys(array_filter($all_rounds, 'is_formatting_round'));
+        // default R_round is first R-round
+        $this->default_R_round_id = $this->R_round_options[0];
+        // default L_round is the preceding one
+        $this->default_L_round_id = $all_ids[array_search($this->default_R_round_id, $all_ids) - 1];
```

Comment:
This might be easier if you use the objects rather than the IDs for all of these and then just use `get_round_param()` to get the object. We probably don't really care if they manipulate the URL and get something they weren't expecting.
```php
       // L_round_options are all rounds except the last
       $this->L_round_options = array_slice($all_rounds, 0, count($all_rounds) - 1);
       // R_round_options are all formatting rounds
       $this->R_round_options = array_filter($all_rounds, 'is_formatting_round');
       // default R_round is first R-round
       $this->default_R_round = $this->R_round_options[0];
       // default L_round is the preceding one
       $this->default_L_round = Rounds::get_by_number($this->default_R_round->round_number - 1);

       $this->L_round_id = get_round_param($_GET, "L_round_id", $this->default_L_round)->id;
       $this->R_round_id = get_round_param($_GET, "R_round_id", $this->default_R_round)->id;
```
Although looking at the rest of the script it might be clearer if the whole script dealt with round objects instead of round IDs. Because ultimately we convert them into round objects anyway:
```php
        $L_round = get_Round_for_round_id($this->L_round_id);
        $R_round = get_Round_for_round_id($this->R_round_id);
```

Reply:
I've done this and removed some of the member functions so the $L_round and $r_round do not have to be class members.
It would be possible to remove the class altogether if we put some parameters in the functions. It seems a bit OTT to use a class just to avoid a few parameters or global variables.